libass/ass_shaper.c

```diff
@@ -231,7 +231,7 @@ get_cached_metrics(struct ass_shaper_metrics_data *metrics,
     bool rotate = false;
     // if @font rendering is enabled and the glyph should be rotated,
     // make cached_h_advance pick up the right advance later
-    if (metrics->vertical && unicode >= VERTICAL_LOWER_BOUND)
+    if (metrics->vertical && ass_codepoint_is_fullwidth(metrics->hash_key.font->faces_cp[metrics->hash_key.face_index], unicode))
```

Comment:
I fear this can be really slow. Maybe it's better to move this check into metrics cache?

Reply:
Sure; done.